[FrameworkBundle] Remove deprecate session service

jderusse opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.0
| Bug fix?      | no
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

To ease code review this PR only removes code for session. 

Blocked by #41343

jderusse committed:
Remove deprecate session service

nicolas-grekas commented:
Rebase unlocked

fabpot commented:
@jderusse Is it still in draft mode?

nicolas-grekas commented on src/Symfony/Component/HttpFoundation/Session/Storage/ServiceSessionFactory.php:
```diff
@@ -16,7 +16,7 @@
  * @author Jérémy Derussé <jeremy@derusse.com>
  *
- * @internal to be removed in Symfony 6
+ * @internal to be removed when bumping to symfony/framework-bundle:6.0
```
what does this mean?

jderusse replied:
`symfony/framework-bundle: 5.3 to 5.4.*` uses this class. 
We can delete it when `symfony/http-foundation` will conflict with `symfony/framework-bundle < 6`

jderusse replied:
classe removed

stof replied:
is FrameworkBundle 5.3 allowing symfony/http-foundation 6 ? that would be bad as it was not released yet at that time.

And to solcve that, I'd rather make FrameworkBundle 5.4 require symfony/http-foundation 5.4 only instead of allowing 6.x as well. Otherwise, once we add a conflict rule in symfony/http-foundation 6.1, FrameworkBundle 5.4 LTS would end up installing symfony/http-foundation `5.4 || 6.0` which will install 6.0 which will be unmaintained a lot before the end of the LTS.

nicolas-grekas replied:
5.3 no, but 5.4 yes, and we should try to keep that for as long as possible, to put our own BC layer into practice
flex is the solution to stick to 5.4 for transitive deps.

jderusse commented:
:green_heart: green

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/AbstractSessionListener.php:
```diff
@@ -62,9 +61,6 @@ public function onKernelRequest(RequestEvent $event)
         }
-
-        $session = $this->container && $this->container->has('initialized_session') ? $this->container->get('initialized_session') : null;
-        $this->sessionUsageStack[] = $session instanceof Session ? $session->getUsageIndex() : 0;
```
This was supposed to account for subrequests.
I'm not 100% sure to get how usage tracking will work without these lines.
Can you give me some hints about that?

jderusse replied:
To be honest, I don't really understand the purpose of this variable.

This line of code is reached when a `RequestEvent` is dispatched for a `MainRequest` only (see line 54). So, this "stack" should always contain 1 single element (unless I missed something and we can start several `MainRequest` without dispatching a `FinishRequestEvent`)

If sessions are removed from the DIC, they are not shared: every MainRequest should start with a new Fresh Session (usageIndex=0).

If this is a bug:
- Should we fix this upstream?
- Checking the `usageIndex` of the subRequest in the `onKernelRequest` method might break lazyness should we take care of this?

nicolas-grekas replied:
> start several MainRequest without dispatching a FinishRequestEvent

Hum, maybe this is not for subrequests, but for ESI requests, (which are all main requests).
This needs some investigation :)

jderusse replied:
When called by a Proxy with surrogateCapabilities, ESI are 2 distinguished HTTP requests. Moreover, the second request starts after the first one is finished.
When called without surrogateCapabilities, Symfony replaces the proxy (see `AbstractSurrogateFragmentRenderer::render`) use the `InlineFragmentRender` to create subRequest.

As far as I know, there is no case with 2 nested MainRequest

jderusse replied:
I read again this class (and related PR https://github.com/symfony/symfony/pull/27467) and I definitely think this variable didn't have to be an array at first. There is no case where we could have 2 nested MainRequests.

Given in 6.0 the 'session' service has been removed, each request have a fresh new Session, we don't have to remember the previous value of the the usageIndex.

nicolas-grekas replied:
Ok, I don't either. Works for me.

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/AbstractTestSessionListener.php:
```diff
@@ -102,17 +100,8 @@ public function onKernelResponse(ResponseEvent $event)
     {
         return [
-            KernelEvents::REQUEST => ['onKernelRequest', 127], // AFTER SessionListener
+            KernelEvents::REQUEST => ['onKernelRequest', 127], // After SessionListener
```
needless merge conflict ahead :)

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/SessionListener.php:
```diff
@@ -34,29 +34,8 @@ public function __construct(ContainerInterface $container, bool $debug = false)
-            && ($mainRequest = $this->container->get('request_stack')->getMainRequest())
-            && $mainRequest->isSecure()
-        ) {
-            $storage->setOptions(['cookie_secure' => true]);
```
the phpdoc on the class needs an update

nicolas-grekas commented on src/Symfony/Component/Security/Http/Tests/Firewall/ContextListenerTest.php:
```diff
@@ -413,17 +413,10 @@ private function handleEventWithPreviousSession($userProviders, UserInterface $u
-                },
-            ]
-        ) implements ContainerInterface {
+        $tokenStorage = new UsageTrackingTokenStorage($tokenStorage, new class(['request_stack' => function () use ($requestStack) {
```
Maybe split the line here to move the key on a dedicated one?

nicolas-grekas commented on src/Symfony/Component/HttpKernel/EventListener/AbstractSessionListener.php:
```diff
@@ -78,9 +74,10 @@ public function onKernelResponse(ResponseEvent $event)
         // Always remove the internal header if present
         $response->headers->remove(self::NO_AUTO_CACHE_CONTROL_HEADER);
 
-        if (!$session = $this->container && $this->container->has('initialized_session') ? $this->container->get('initialized_session') : $event->getRequest()->getSession()) {
```
Can you shed my light on why we don't need `initialized_session` anymore?
IIRC, it was there to skip accessing the session when it's not been used before.

jderusse replied:
Added a local property to track usage of the Factory. That prevent calling `$request->getSession()` for nothing in the `onKernelResponse` method.

Beware. This also breaks the behavior if the user sets its own session in the request (which is already the case in symfony < 6 that use `initialized_session`).

- onKernelRequest => the factory is injected in the request
- user call `request->setSession(new Session())`
- onKernelResponse realize that the factory has not been called => early exit

If we want an accurate way to check if the session has been warmd, we should add a method in the request `hasInitializedSession` or `hasSessionFactory` or `hasSession(bool $ignoreFactory = false)`

nicolas-grekas replied:
WDYT of `hasSession(bool $skipIfUninitialized = false)`, with a BC layer on 5.4?

jderusse replied:
See #42900

jderusse committed:
Fix tests when security-http 5.4 uses security-core 6.0

nicolas-grekas committed:
minor #42907 Fix tests when security-http 5.4 uses security-core 6.0 (jderusse)

This PR was merged into the 5.4 branch.

Discussion
----------

Fix tests when security-http 5.4 uses security-core 6.0

| Q             | A
| ------------- | ---
| Branch?       | 5.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no
| Tickets       | -
| License       | MIT
| Doc PR        | -

see falling test in flipped tests for 6.0 branch: https://github.com/symfony/symfony/pull/41321/checks?check_run_id=3528227747#step:8:3572

Commits
-------

58f09e55fb Fix tests when security-http 5.4 uses security-core 6.0

nicolas-grekas committed:
cs fix

nicolas-grekas committed:
Merge branch '5.4' into 6.0

* 5.4:
  cs fix
  Fix tests when security-http 5.4 uses security-core 6.0

nicolas-grekas commented:
Thank you @jderusse.
